src/poprox_recommender/pipeline.py

```diff
@@ -0,0 +1,56 @@
+from copy import deepcopy
+from dataclasses import dataclass
+from typing import Callable
+
+from poprox_concepts import ArticleSet, InterestProfile
+
+
+@dataclass
+class ComponentSpec:
+    component: Callable
+    inputs: list[str]
+    output: str
+
+
+class RecommendationPipeline:
+    def __init__(self, name):
+        self.name = name
+        self.components = []
+
+    def add(self, component: Callable, inputs: list[str], output: str):
+        self.components.append(ComponentSpec(component, inputs, output))
+
+    def __call__(self, inputs: dict[str, ArticleSet | InterestProfile]) -> ArticleSet:
+        # Avoid modifying the inputs
+        state = deepcopy(inputs)
+
+        # Run each component in the order it was added
+        for component_spec in self.components:
+            state = self.run_component(component_spec, state)
+
+        recs = state[self.components[-1].output]
+
+        # Double check that we're returning the right type for recs
+        if not isinstance(recs, ArticleSet):
+            msg = f"The final pipeline component must return ArticleSet, but received {type(recs)}"
+            raise TypeError(msg)
+
+        return recs
+
+    def run_component(self, component_spec: ComponentSpec, state: dict[str, ArticleSet | InterestProfile]):
+        arguments = []
+        for input_name in component_spec.inputs:
+            arguments.append(state[input_name])
```

Comment:
In theory we should validate the types of the inputs and raise `TypeError`s here but I suspect that might be code we'll just rip back out. I didn't hit any cases where clearer error messages on inputs would have helped me, but I could be persuaded to add them if it seems useful.

Reply:
his was the quickest, cheapest pipeline I could come up with so my main thought was "this is a smaller, simpler step forward," but I wouldn't be opposed to moving toward keyword arguments, better type-checking, etc